docs/vm/proposal.md

```diff
@@ -0,0 +1,562 @@
+
+The following will be true at the end of this phase:
+
+* IR is generated for native flux functions if all operations are supported.
+* Recursive function definitions are generated even if the function they are contained within is not.
+* Generics are implemented.
+* Function objects exist.
+* Closures are not implemented.
+* Indirect call instruction is implemented in the IR and capable of reifying functions.
+* Imports and packages are not implemented.
+
+### Phase 3
+
+Phase 3 is focused on records, their data layout, passing them as arguments, and accessing them.
+
+A primary requirement of this phase is that record literals should be cheap to allocate and use.
+In the current implementation, record literals create a memory allocation.
+In this version, records should not allocate memory outside the VM stack.
+
+The following will be true at the end of this phase:
+
+* Record literals can be used and returned from functions.
+* Records can be passed as an argument to a function.
+* Records will be representable directly on the stack.
+* Member expressions will be implemented in the IR and VM.
+
+### Phase 4
+
+Phase 4 is focused on implementing closures.
+
+The following will be true at the end of this phase:
+
+* Values that are closed over will be saved as part of the function object.
+* Closure mechanics can be visibly seen from the IR.
+
+### Phase 5
+
+Phase 5 is focused on importing packages and using builtin functions that are not streams.
+
+The following will be true at the end of this phase:
+
+* Imports function.
+* Builtin values that do not interact with table streams will work.
+
+At this point, the remaining instructions can be implemented separately and the VM can replace the compiler for most circumstances.
+This includes any remaining instructions that have not been implemented from phase 1.
+
+### Phase 6
+
+Phase 6 is focused on integrating table streams into the VM.
+Functions that return table streams will construct a plan and the VM will optimize the plan and then pass it to the executor/dispatcher.
+
+The following will be true at the end of this phase:
+
+* Sources and transformations are constructed by the VM.
+* Constructed plans are passed through the planner and executed by the executor.
+* The yield instruction is implemented and is capable of returning table stream data.
+* Table find style functions do not work.
+
+### Phase 7
+
+Phase 7 is focused on ensuring table find style functions work properly within the VM.
+
+The following will be true at the end of this phase:
+
+* Table find style functions will be capable of using the same planner/executor as yielded table streams.
+* Table find style functions work.
+
+### Phase 8
+
+Phase 8 is focused on polishing up the implementation and ensuring it is feature complete.
+This phase can be started as early as after phase 5 and is primarily focused on implementing any remaining instructions from the semantic graph that haven't been implemented.
```

Comment:
This can be done concurrently and by different people. I just wanted to focus on the stream portion because it's so important that it would need to be prioritized. If the stream work doesn't succeed, phase 8 is mostly useless because we can't finish the project. I do think that after phase 5 we can start dividing the work between more people to do exactly what you mention.

At this point in the project, the VM would already be inserted as an alternative to the compiler so I don't think there's more integration work.